local-tests/chainrunner/chain.py

```diff
@@ -162,3 +157,57 @@ def purge(self, nodes=None):
         idx = nodes or range(len(self.nodes))
         for i in idx:
             self.nodes[i].purge()
+
+    def fork(self, forkoff_path, ws_endpoint):
+        """Replace the chainspec of this chain with the state forked from the given `ws_endpoint`.
+        This method should be run after bootstrapping the chain, but before starting it.
+        'forkoff_path' should be a path to fork-off binary."""
+        forked = op.join(self.path, 'forked.json')
+        cmd = [check_file(forkoff_path), '--ws-rpc-endpoint', ws_endpoint,
+                '--initial-spec-path', op.join(self.path, 'chainspec.json'),
+                '--snapshot-path', op.join(self.path, 'snapshot.json'),
+                '--combined-spec-path', forked]
+        subprocess.run(cmd, check=True)
+        self.set_chainspec(forked)
+
+    def update_runtime(self, cliain_path, sudo_phrase, runtime):
+        """Send set_code extrinsic with runtime update.
+        Requires a path to `cliain` binary, a path to new WASM runtime and the sudo phrase."""
+        port = self.nodes[0].ws_port()
+        cmd = [check_file(cliain_path), '--node', f'localhost:{port}', '--seed', sudo_phrase,
+                'update-runtime', '--runtime', check_file(runtime)]
+        subprocess.run(cmd, check=True)
+
+    def wait_for_finalization(self, old_finalized, nodes=None, timeout=300, finalized_delta=3, catchup=True, catchup_delta=10):
+        """Wait for finalization to catch up with the newest blocks. Requires providing the number
+        of the last finalized block, which will be used as a reference against recently finalized blocks.
+        The finalization is considered "recovered" when all provided `nodes` (all nodes if None)
+        have seen a finalized block higher than `old_finalized` + `finalized_delta`.
+        If `catchup` is True, wait until finalization catches up with the newly produced blocks
+        (within `catchup_delta` blocks). 'timeout' (in seconds) is a global timeout for the whole method
+        to execute. Raise TimeoutError if finalization fails to recover within the given timeout."""
+        nodes = [self.nodes[i] for i in nodes] if nodes else self.nodes
+        deadline = time.time() + timeout
+        while any((n.highest_block()[1] <= old_finalized + finalized_delta) for n in nodes):
+            time.sleep(5)
+            if time.time() > deadline:
+                raise TimeoutError(f'Block finalization stalled after {timeout} seconds')
+        if catchup:
+            def lags(node):
+                r, f = node.highest_block()
+                return r - f > catchup_delta
```

Comment:
what about `lags = lambda node: operator.sub(node.highest_block())`?

Reply:
Much less readable in my opinion. And requires new import ;)